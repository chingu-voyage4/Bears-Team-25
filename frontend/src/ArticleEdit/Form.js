import React from 'react';
import PropTypes from 'prop-types';

import Typography from '@material-ui/core/Typography';
import MenuItem from '@material-ui/core/MenuItem';
import List from '@material-ui/core/List';
import ListItem from '@material-ui/core/ListItem';

import ListItemInput from '../ListItemInput';
import TopicSelector from '../TopicEdit/TopicSelector';

import { Label } from './styled';

/* eslint-disable camelcase */
export default class Form extends React.Component {
  static propTypes = {
    mobile: PropTypes.bool.isRequired,
    title: PropTypes.string.isRequired,
    order: PropTypes.string.isRequired,
    topics: PropTypes.array.isRequired,
    sub_topics: PropTypes.array.isRequired,
    selectedTopic: PropTypes.object.isRequired,
    selectedSubTopic: PropTypes.any,
    handleFieldChange: PropTypes.func.isRequired,
    setTopics: PropTypes.func.isRequired,
  };

  static defaultProps = {
    selectedSubTopic: Form.nullSubTopic,
  };

  static nullSubTopic = { _id: '0' };

  handleTopicSelect = e => {
    const selectedTopic = this.props.topics.reduce((acc, topic) => {
      if (topic._id === e.target.value) {
        return topic;
      }
      return acc;
    }, {});
    const selectedSubTopic = null;
    this.props.setTopics(selectedTopic, selectedSubTopic);
  };

  handleSubTopicSelect = e => {
    const selectedSubTopic = this.props.sub_topics.reduce((acc, sub) => {
      if (sub._id === e.target.value) {
        return sub;
      }
      return acc;
    }, null);
    this.props.setTopics(this.props.selectedTopic, selectedSubTopic);
  };
  render() {
    const {
      mobile,
      title,
      order,
      handleFieldChange,
      topics,
      sub_topics,
      selectedTopic,
      selectedSubTopic,
    } = this.props;
    const topicList = topics.map(topic => (
      <MenuItem key={topic._id} value={topic._id}>
        {topic.name}
      </MenuItem>
    ));
    const subTopicList = sub_topics.reduce((acc, sub) => {
      let cac = [...acc];
      if (acc.length === 0) {
        cac = acc.concat(
          <MenuItem key={0} value="0">
            None
          </MenuItem>
        );
      }
      if (sub.parent === selectedTopic._id) {
        return cac.concat(
          <MenuItem key={sub._id} value={sub._id}>
            {sub.name}
          </MenuItem>
        );
      }
      return acc;
    }, []);
    const fixedSelectedSubTopic = selectedSubTopic || Form.nullSubTopic;
    return (
      <List>
        <ListItemInput
          mobile={mobile}
          label="Title"
          name="title"
          value={title}
          onChange={handleFieldChange}
        />
        <ListItemInput
          mobile={mobile}
          label="Order"
          name="order"
          value={order}
          onChange={handleFieldChange}
        />
        <ListItem>
          <Label>
            <Typography variant="title">Topic:</Typography>
          </Label>
          <TopicSelector
            selectedTopic={selectedTopic._id}
            onSelect={this.handleTopicSelect}
            topicList={topicList}
          />
        </ListItem>
        <ListItem>
          <Label>
            <Typography variant="title">Sub Topic:</Typography>
          </Label>
          <TopicSelector
            selectedTopic={fixedSelectedSubTopic._id}
            onSelect={this.handleSubTopicSelect}
            topicList={subTopicList}
          />
        </ListItem>
      </List>
    );
  }
}
