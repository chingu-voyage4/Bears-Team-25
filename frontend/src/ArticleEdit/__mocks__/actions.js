/* eslint-disable import/prefer-default-export */
import article from './articleData';

// eslint-disable-next-line no-unused-vars
const save = payload =>
  new Promise(resolve => resolve({ success: true, _id: article._id }));

const get = id =>
  new Promise((resolve, reject) => {
    if (id === '43039ac8d0a244719d9d31e0731bcbe8') {
      resolve(article);
    }
    return reject(new Error('article id not found'));
  });

export default { save, get };
