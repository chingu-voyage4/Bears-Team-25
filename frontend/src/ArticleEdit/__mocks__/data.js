export const topics = [
  { _id: '1', name: 'Voyage', order: 1 },
  { _id: '2', name: 'PMRoK', order: 2 },
];
export const subTopics = [
  { _id: '10', parent: '1', name: 'About wiki', order: 1 },
  { _id: '11', parent: '1', name: 'About Voyages', order: 2 },
  { _id: '12', parent: '1', name: 'About PMRoK', order: 1 },
];
export const article = {
  _id: '43039ac8d0a244719d9d31e0731bcbe8',
  active: true,
  created: '1498771528',
  creator: {
    _id: '85e026bbd79b46ce990dbb956490e8d0',
    username: 'viktor',
  },
  title: 'About this wiki',
  order: 1,
  topic: topics[0],
  sub_topic: subTopics[0],
  content: 'test content',
};
