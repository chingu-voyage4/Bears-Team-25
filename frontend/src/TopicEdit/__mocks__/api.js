/* eslint-disable import/prefer-default-export */

export const getTopics = () =>
  Promise.resolve([{ _id: '1', name: 'Voyage', order: 1 }]);

export const getSubTopics = () =>
  Promise.resolve(
    // eslint-disable-next-line camelcase
    [{ _id: '2', parent: '1', name: 'About this wiki', order: 1 }]
  );
