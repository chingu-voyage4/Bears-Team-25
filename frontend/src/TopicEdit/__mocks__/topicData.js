export const topics = [
  { _id: '1', name: 'Voyage', order: 1 },
  { _id: '2', name: 'PMRoK', order: 2 },
];
export const subTopics = [
  { _id: '10', parent: '1', name: 'About wiki', order: 1 },
  { _id: '11', parent: '1', name: 'About Voyages', order: 2 },
  { _id: '12', parent: '1', name: 'About PMRoK', order: 1 },
];
