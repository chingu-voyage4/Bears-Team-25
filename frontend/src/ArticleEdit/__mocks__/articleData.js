import { topics, subTopics } from '../../TopicEdit/__mocks__/topicData';

const article = {
  _id: '43039ac8d0a244719d9d31e0731bcbe8',
  active: true,
  created: '1498771528',
  creator: {
    _id: '85e026bbd79b46ce990dbb956490e8d0',
    username: 'viktor',
  },
  title: 'About this wiki',
  order: 1,
  topic: topics[0],
  sub_topic: subTopics[0],
  content: 'test content',
  edit_lock: false,
};

export default article;
