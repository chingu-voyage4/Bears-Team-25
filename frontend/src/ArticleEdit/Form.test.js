import React from 'react';
import renderer from 'react-test-renderer';
import { shallow, mount } from 'enzyme';

import { topics, subTopics } from '../TopicEdit/__mocks__/topicData';
import Form from './Form';

it('renders ArticleEdit Form', () => {
  const handleFieldChange = jest.fn();
  const setTopics = jest.fn();
  const comp = renderer.create(
    <Form
      mobile={false}
      title="title"
      order="1"
      topics={topics}
      sub_topics={subTopics}
      selectedTopic={topics[0]}
      selectedSubTopic={subTopics[0]}
      handleFieldChange={handleFieldChange}
      setTopics={setTopics}
    />
  );
  const tree = comp.toJSON();
  expect(tree).toMatchSnapshot();
});

it('renders ArticleEdit Form', () => {
  const handleFieldChange = jest.fn();
  const setTopics = jest.fn();
  const comp = renderer.create(
    <Form
      mobile={false}
      title="title"
      order="1"
      topics={topics}
      sub_topics={subTopics}
      selectedTopic={topics[0]}
      selectedSubTopic={null}
      handleFieldChange={handleFieldChange}
      setTopics={setTopics}
    />
  );
  const tree = comp.toJSON();
  expect(tree).toMatchSnapshot();
});

describe('Form ops', () => {
  let comp;
  let handleFieldChange;
  let setTopics;
  let inst;
  beforeEach(() => {
    handleFieldChange = jest.fn();
    setTopics = jest.fn();
    comp = shallow(
      <Form
        mobile={false}
        title="title"
        order="1"
        topics={topics}
        sub_topics={subTopics}
        selectedTopic={topics[0]}
        selectedSubTopic={subTopics[0]}
        handleFieldChange={handleFieldChange}
        setTopics={setTopics}
      />
    );
    inst = comp.instance();
  });
  it('handles change of topic', () => {
    inst.handleTopicSelect({ target: { value: topics[1]._id } });
    expect(setTopics).toHaveBeenCalledWith(topics[1], null);
  });
  it('handles change of sub-topic', () => {
    inst.handleSubTopicSelect({ target: { value: subTopics[0]._id } });
    expect(setTopics).toHaveBeenCalledWith(topics[0], subTopics[0]);
  });
});
