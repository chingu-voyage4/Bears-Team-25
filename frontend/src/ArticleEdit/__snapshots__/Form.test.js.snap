// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`renders ArticleEdit Form passed a null sub-topic 1`] = `
<ul
  className="MuiList-root-1 MuiList-padding-2"
>
  <li
    className="MuiListItem-root-5 MuiListItem-default-8 MuiListItem-gutters-12"
    disabled={false}
  >
    <div
      className="sc-bdVaJa drcPZX"
    >
      <h2
        className="MuiTypography-root-15 MuiTypography-title-21"
      >
        Title
        :
      </h2>
    </div>
    <div
      className="MuiListItemText-root-39"
    >
      <h3
        className="MuiTypography-root-15 MuiTypography-subheading-22 MuiListItemText-primary-42"
      >
        <div
          aria-describedby={undefined}
          className="MuiFormControl-root-45 MuiFormControl-fullWidth-48"
          onBlur={[Function]}
          onFocus={[Function]}
        >
          <div
            className="MuiInput-root-49 MuiInput-fullWidth-56 MuiInput-formControl-50 MuiInput-underline-53"
          >
            <input
              aria-invalid={false}
              aria-required={false}
              autoComplete={undefined}
              autoFocus={undefined}
              className="MuiInput-input-57"
              defaultValue={undefined}
              disabled={false}
              id={undefined}
              name="title"
              onBlur={[Function]}
              onChange={[Function]}
              onFocus={[Function]}
              onKeyDown={undefined}
              onKeyUp={undefined}
              placeholder={undefined}
              readOnly={undefined}
              required={undefined}
              rows={undefined}
              type="text"
              value="title"
            />
          </div>
        </div>
      </h3>
    </div>
  </li>
  <li
    className="MuiListItem-root-5 MuiListItem-default-8 MuiListItem-gutters-12"
    disabled={false}
  >
    <div
      className="sc-bdVaJa drcPZX"
    >
      <h2
        className="MuiTypography-root-15 MuiTypography-title-21"
      >
        Order
        :
      </h2>
    </div>
    <div
      className="MuiListItemText-root-39"
    >
      <h3
        className="MuiTypography-root-15 MuiTypography-subheading-22 MuiListItemText-primary-42"
      >
        <div
          aria-describedby={undefined}
          className="MuiFormControl-root-45 MuiFormControl-fullWidth-48"
          onBlur={[Function]}
          onFocus={[Function]}
        >
          <div
            className="MuiInput-root-49 MuiInput-fullWidth-56 MuiInput-formControl-50 MuiInput-underline-53"
          >
            <input
              aria-invalid={false}
              aria-required={false}
              autoComplete={undefined}
              autoFocus={undefined}
              className="MuiInput-input-57"
              defaultValue={undefined}
              disabled={false}
              id={undefined}
              name="order"
              onBlur={[Function]}
              onChange={[Function]}
              onFocus={[Function]}
              onKeyDown={undefined}
              onKeyUp={undefined}
              placeholder={undefined}
              readOnly={undefined}
              required={undefined}
              rows={undefined}
              type="text"
              value="1"
            />
          </div>
        </div>
      </h3>
    </div>
  </li>
  <li
    className="MuiListItem-root-5 MuiListItem-default-8 MuiListItem-gutters-12"
    disabled={false}
  >
    <div
      className="sc-iwsKbI jmYMFI"
    >
      <h2
        className="MuiTypography-root-15 MuiTypography-title-21"
      >
        Topic:
      </h2>
    </div>
    <div
      className="MuiInput-root-49 MuiInput-underline-53 sc-ifAKCX dUasrT"
    >
      <div
        className="MuiSelect-root-63"
      >
        <div
          aria-haspopup="true"
          aria-owns={null}
          aria-pressed="false"
          className="MuiSelect-select-64 MuiSelect-selectMenu-65 MuiInput-input-57"
          onBlur={[Function]}
          onClick={[Function]}
          onFocus={[Function]}
          onKeyDown={[Function]}
          role="button"
          tabIndex={0}
        >
          Voyage
        </div>
        <input
          aria-invalid={undefined}
          aria-required={undefined}
          autoComplete={undefined}
          autoFocus={undefined}
          defaultValue={undefined}
          id={undefined}
          name={undefined}
          onKeyDown={undefined}
          onKeyUp={undefined}
          placeholder={undefined}
          readOnly={undefined}
          required={undefined}
          rows={undefined}
          type="hidden"
          value="1"
        />
        <svg
          aria-hidden="true"
          className="MuiSvgIcon-root-68 MuiSelect-icon-67"
          color={undefined}
          focusable="false"
          viewBox="0 0 24 24"
        >
          <path
            d="M7 10l5 5 5-5z"
          />
        </svg>
      </div>
    </div>
  </li>
  <li
    className="MuiListItem-root-5 MuiListItem-default-8 MuiListItem-gutters-12"
    disabled={false}
  >
    <div
      className="sc-iwsKbI jmYMFI"
    >
      <h2
        className="MuiTypography-root-15 MuiTypography-title-21"
      >
        Sub Topic:
      </h2>
    </div>
    <div
      className="MuiInput-root-49 MuiInput-underline-53 sc-ifAKCX dUasrT"
    >
      <div
        className="MuiSelect-root-63"
      >
        <div
          aria-haspopup="true"
          aria-owns={null}
          aria-pressed="false"
          className="MuiSelect-select-64 MuiSelect-selectMenu-65 MuiInput-input-57"
          onBlur={[Function]}
          onClick={[Function]}
          onFocus={[Function]}
          onKeyDown={[Function]}
          role="button"
          tabIndex={0}
        >
          None
        </div>
        <input
          aria-invalid={undefined}
          aria-required={undefined}
          autoComplete={undefined}
          autoFocus={undefined}
          defaultValue={undefined}
          id={undefined}
          name={undefined}
          onKeyDown={undefined}
          onKeyUp={undefined}
          placeholder={undefined}
          readOnly={undefined}
          required={undefined}
          rows={undefined}
          type="hidden"
          value="0"
        />
        <svg
          aria-hidden="true"
          className="MuiSvgIcon-root-68 MuiSelect-icon-67"
          color={undefined}
          focusable="false"
          viewBox="0 0 24 24"
        >
          <path
            d="M7 10l5 5 5-5z"
          />
        </svg>
      </div>
    </div>
  </li>
</ul>
`;

exports[`renders ArticleEdit Form with sub-topic 1`] = `
<ul
  className="MuiList-root-1 MuiList-padding-2"
>
  <li
    className="MuiListItem-root-5 MuiListItem-default-8 MuiListItem-gutters-12"
    disabled={false}
  >
    <div
      className="sc-bdVaJa drcPZX"
    >
      <h2
        className="MuiTypography-root-15 MuiTypography-title-21"
      >
        Title
        :
      </h2>
    </div>
    <div
      className="MuiListItemText-root-39"
    >
      <h3
        className="MuiTypography-root-15 MuiTypography-subheading-22 MuiListItemText-primary-42"
      >
        <div
          aria-describedby={undefined}
          className="MuiFormControl-root-45 MuiFormControl-fullWidth-48"
          onBlur={[Function]}
          onFocus={[Function]}
        >
          <div
            className="MuiInput-root-49 MuiInput-fullWidth-56 MuiInput-formControl-50 MuiInput-underline-53"
          >
            <input
              aria-invalid={false}
              aria-required={false}
              autoComplete={undefined}
              autoFocus={undefined}
              className="MuiInput-input-57"
              defaultValue={undefined}
              disabled={false}
              id={undefined}
              name="title"
              onBlur={[Function]}
              onChange={[Function]}
              onFocus={[Function]}
              onKeyDown={undefined}
              onKeyUp={undefined}
              placeholder={undefined}
              readOnly={undefined}
              required={undefined}
              rows={undefined}
              type="text"
              value="title"
            />
          </div>
        </div>
      </h3>
    </div>
  </li>
  <li
    className="MuiListItem-root-5 MuiListItem-default-8 MuiListItem-gutters-12"
    disabled={false}
  >
    <div
      className="sc-bdVaJa drcPZX"
    >
      <h2
        className="MuiTypography-root-15 MuiTypography-title-21"
      >
        Order
        :
      </h2>
    </div>
    <div
      className="MuiListItemText-root-39"
    >
      <h3
        className="MuiTypography-root-15 MuiTypography-subheading-22 MuiListItemText-primary-42"
      >
        <div
          aria-describedby={undefined}
          className="MuiFormControl-root-45 MuiFormControl-fullWidth-48"
          onBlur={[Function]}
          onFocus={[Function]}
        >
          <div
            className="MuiInput-root-49 MuiInput-fullWidth-56 MuiInput-formControl-50 MuiInput-underline-53"
          >
            <input
              aria-invalid={false}
              aria-required={false}
              autoComplete={undefined}
              autoFocus={undefined}
              className="MuiInput-input-57"
              defaultValue={undefined}
              disabled={false}
              id={undefined}
              name="order"
              onBlur={[Function]}
              onChange={[Function]}
              onFocus={[Function]}
              onKeyDown={undefined}
              onKeyUp={undefined}
              placeholder={undefined}
              readOnly={undefined}
              required={undefined}
              rows={undefined}
              type="text"
              value="1"
            />
          </div>
        </div>
      </h3>
    </div>
  </li>
  <li
    className="MuiListItem-root-5 MuiListItem-default-8 MuiListItem-gutters-12"
    disabled={false}
  >
    <div
      className="sc-iwsKbI jmYMFI"
    >
      <h2
        className="MuiTypography-root-15 MuiTypography-title-21"
      >
        Topic:
      </h2>
    </div>
    <div
      className="MuiInput-root-49 MuiInput-underline-53 sc-ifAKCX dUasrT"
    >
      <div
        className="MuiSelect-root-63"
      >
        <div
          aria-haspopup="true"
          aria-owns={null}
          aria-pressed="false"
          className="MuiSelect-select-64 MuiSelect-selectMenu-65 MuiInput-input-57"
          onBlur={[Function]}
          onClick={[Function]}
          onFocus={[Function]}
          onKeyDown={[Function]}
          role="button"
          tabIndex={0}
        >
          Voyage
        </div>
        <input
          aria-invalid={undefined}
          aria-required={undefined}
          autoComplete={undefined}
          autoFocus={undefined}
          defaultValue={undefined}
          id={undefined}
          name={undefined}
          onKeyDown={undefined}
          onKeyUp={undefined}
          placeholder={undefined}
          readOnly={undefined}
          required={undefined}
          rows={undefined}
          type="hidden"
          value="1"
        />
        <svg
          aria-hidden="true"
          className="MuiSvgIcon-root-68 MuiSelect-icon-67"
          color={undefined}
          focusable="false"
          viewBox="0 0 24 24"
        >
          <path
            d="M7 10l5 5 5-5z"
          />
        </svg>
      </div>
    </div>
  </li>
  <li
    className="MuiListItem-root-5 MuiListItem-default-8 MuiListItem-gutters-12"
    disabled={false}
  >
    <div
      className="sc-iwsKbI jmYMFI"
    >
      <h2
        className="MuiTypography-root-15 MuiTypography-title-21"
      >
        Sub Topic:
      </h2>
    </div>
    <div
      className="MuiInput-root-49 MuiInput-underline-53 sc-ifAKCX dUasrT"
    >
      <div
        className="MuiSelect-root-63"
      >
        <div
          aria-haspopup="true"
          aria-owns={null}
          aria-pressed="false"
          className="MuiSelect-select-64 MuiSelect-selectMenu-65 MuiInput-input-57"
          onBlur={[Function]}
          onClick={[Function]}
          onFocus={[Function]}
          onKeyDown={[Function]}
          role="button"
          tabIndex={0}
        >
          About wiki
        </div>
        <input
          aria-invalid={undefined}
          aria-required={undefined}
          autoComplete={undefined}
          autoFocus={undefined}
          defaultValue={undefined}
          id={undefined}
          name={undefined}
          onKeyDown={undefined}
          onKeyUp={undefined}
          placeholder={undefined}
          readOnly={undefined}
          required={undefined}
          rows={undefined}
          type="hidden"
          value="10"
        />
        <svg
          aria-hidden="true"
          className="MuiSvgIcon-root-68 MuiSelect-icon-67"
          color={undefined}
          focusable="false"
          viewBox="0 0 24 24"
        >
          <path
            d="M7 10l5 5 5-5z"
          />
        </svg>
      </div>
    </div>
  </li>
</ul>
`;
